frame/contracts/COMPLEXITY.md

```diff
@@ -262,19 +260,40 @@ Also, `transfer` can make up to 2 DB reads and up to 2 DB writes (if flushed to
 
 Storing the code in the overlay may induce another DB write (if flushed to the storage) with the size proportional to the size of the constructor code.
 
-Finally, all changes are `commit`-ted into the underlying overlay. The complexity of this depends on the number of changes performed by the constructor code. Thus, the pricing of storage modification should account for that.
+Finally, the current storage transaction is closed.. The complexity of this depends on the number of changes performed by the constructor code. Thus, the pricing of storage modification should account for that.
 
 **complexity**: Up to 2 DB reads and induces up to 3 DB writes (if flushed to the storage), one of which is dependent on the size of the code. Additionally, if the source account removal takes place a DB write will be performed per one storage entry that the account has.
 
-# Externalities
+## Contracts API
+
+Each API function invoked from a contract can involve some overhead.
+
+## Getter functions
+
+Those are simple getter functions which copy a requested value to contract memory. They
+all have the following two arguments:
+
+- `output_ptr`: Pointer into contract memory where to copy the value.
+- `output_len_ptr`: Pointer into contract memory where the size of the buffer is stored. The size of the copied value is also stored there.
 
-Each external function invoked from a contract can involve some overhead.
+**complexity**: The size of the returned value is constant for a given runtime. Therefore we
+consider its complexity constant even though some of them might involve at most one DB read.
 
-## ext_gas
+This is the list of getters:
 
-**complexity**: This is of constant complexity.
+- ext_caller
+- ext_address
+- ext_gas_price
+- ext_gas_left
+- ext_balance
+- ext_value_transferred
+- ext_now
+- ext_minimum_balance
+- ext_tombstone_deposit
+- ext_rent_allowance
+- ext_block_number
```

Comment:
In practice, AFAIR, some of the functions actually delegate to the runtime. E.g. `block_number` is actually requested from the system module.

I wonder if we should note that clearly here. Moreover, `ext_now` (unlike the block number getter) actually performs a call to a function defined in the `Trait`, thus, can have different performance characteristics. I am not particularly worried about it, just something that we should document (e.g. `type Time: ...; // we expect constant time execution time`

Reply:
I think we can get away with writing that down as blanket assumption for all those getters. It would be madness if something there had a super linear runtime.

Reply:
Added this assumption.